dmake/core.py

```diff
@@ -247,8 +249,21 @@ def activate_service(loaded_files, service_providers, service_dependencies, comm
             children += activate_base(base_variant)
         elif command == 'run':
             children += activate_service_shared_volumes(loaded_files, service_providers, service)
+            # ~hackish: run service depends on test service if we are doing tests
             if common.command in ['test', 'deploy']:
-                children += activate_service(loaded_files, service_providers, service_dependencies, 'test', service)
+                if common.change_detection:
+                    # in change detection mode, don't add "test service" node: only create the link between run and test if the test node exists.
+                    # the services to be tested are either:
+                    # - created directly via graph construction starting on the target services: the ones that changed
+                    # - independently created from "test child service" to "test parent service"
+                    # we can reach here in the middle of the DAG construction (e.g. when multiple services have changed: we fully work one by one sequentially),
+                    # so we don't know yet if the test node will exist at the end or not.
+                    # the link will be created later, in a second global pass in make()
```

Comment:
Maybe be even more explicit in the other code block reference:

```suggestion
                    # the link will be created later, in a second global pass in make() (look for "<second_pass>" below)
```